### PR TITLE
feat: sqlite full text search

### LINKED ISSUE #103: D2 Use the SQLite text search instead of a separate Whoosh search index
* [FTS5](https://www.sqlite.org/fts5.html)
* [Tutorial](https://www.sqlitetutorial.net/sqlite-full-text-search/)

### PR DESCRIPTION
**Pull request by will7200**
Closes #103 

This aims to be a drop in replacement for whoosh. with little to no API change for our customers.

1. Uses FTS4 as FTS5 requires some escaping of the search terms
2. Since the underlying search implementation strips out certain characters out during tokenization, these characters that might be of value to the customer are lost. Added functionality for the mask and filter operations to allows users to pass a callable to control their search better. Operations supported are those of sqlite3 column operators wrapped in pythonic operations.
  Examples:
   ```python
   db.search("lollipop", mask={"product": lambda col: col == 'ZEBRA-2'})
   db.search("lollipop", mask={"product": lambda col: col.like('%ZEBRA-2%')})
   db.search("lollipop", mask={"product": lambda col: col.in_(['ZEBRA-2', 'ZEBRA-1'])})
    ```

### CODE REVIEW

File: bw2data/search/indices.py [hunk 1]
```diff
@@ -51,19 +118,78 @@ def add_dataset(self, ds):
...
+                        ]
+                    ).execute()
 
     def update_dataset(self, ds):
-        writer = self.get().writer()
-        writer.update_document(**self._format_dataset(ds))
-        writer.commit()
+        with self.db.bind_ctx(MODELS):
+            for model in MODELS:
+                model.delete().where(model.code == ds["code"]).execute()
```
Comment (cmutel) [inline bw2data/search/indices.py]: The `code` by itself isn't enough to guarantee uniqueness - the code is normally made up by hashing "core" data attributes like name, location, and unit. However, if I make a copy of a given database, these attributes won't change (and this is convenient - I can then easily find the corresponding dataset in the copy). So we need to search for both `code` and `database` attribute values - the combination of these two is guaranteed to be unique.

File: bw2data/search/indices.py [hunk 2]
```diff
@@ -51,19 +118,78 @@ def add_dataset(self, ds):
...
+                model.insert(**self._format_dataset(ds)).execute()
 
     def delete_dataset(self, ds):
-        index = self.get()
-        index.delete_by_term("code", ds["code"])
+        with self.db.bind_ctx(MODELS):
+            for model in MODELS:
+                model.delete().where(model.code == ds["code"]).execute()
```
Comment (cmutel) [inline bw2data/search/indices.py]: See above.

**Comment by cmutel** [comment]
@will7200 Awesome! One small comment which is an easy fix, and then fix some test failures, and we should be GTG!

**Comment by will7200** [comment]
@cmutel fixed that and found the underlying causing those test failures. The last two remaining windows failures are due to:
```
PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: 'C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\tmpgwrprhc7\\dxbqkfxeszgknljeyz.46382e12\\lci\\databases.db'
```
Which looks like has happened in past github runs.

**Pull request merged**